src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -0,0 +1,141 @@
+﻿using System;
+using System.Linq;
+using System.ComponentModel;
+using System.ComponentModel.Composition;
+using GitHub.Models;
+using GitHub.Logging;
+using Serilog;
+using EnvDTE;
+
+namespace GitHub.Services
+{
+    /// <summary>
+    /// This implementation listenes to IGitExt for ActiveRepositories property change events and fires
+    /// <see cref="PropertyChanged"/> and <see cref="StatusChanged"/> events when appropriate.
+    /// </summary>
+    /// <remarks>
+    /// A <see cref="PropertyChanged"/> is fired when a solution is opened in a new repository (or not in a repository).
+    /// A <see cref="StatusChanged"/> event is only fired when the current branch, head SHA or tracked SHA changes (not
+    /// on every IGitExt property change event). <see cref="ActiveRepository"/> contains the active repository or null
+    /// if a solution is opened that isn't in a repository. No events are fired when the same solution is unloaded then
+    /// reloaded (e.g. when a .sln file is touched).
+    /// </remarks>
+    [Export(typeof(ITeamExplorerContext))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class TeamExplorerContext : ITeamExplorerContext
+    {
+        static ILogger log = LogManager.ForContext<TeamExplorerContext>();
+
+        readonly DTE dte;
+        readonly IVSGitExt gitExt;
+
+        string solutionPath;
+        string repositoryPath;
+        string branchName;
+        string headSha;
+        string trackedSha;
+
+        ILocalRepositoryModel repositoryModel;
+
+        [ImportingConstructor]
+        public TeamExplorerContext(IGitHubServiceProvider serviceProvider, IVSGitExt gitExt)
+            : this(gitExt, serviceProvider)
+        {
+        }
+
+        public TeamExplorerContext(IVSGitExt gitExt, IGitHubServiceProvider serviceProvider)
+        {
+            this.gitExt = gitExt;
+
+            // This is a standard service which should always be available.
+            dte = serviceProvider.GetService<DTE>();
+
+            Refresh();
+            gitExt.ActiveRepositoriesChanged += Refresh;
+        }
+
+        void Refresh()
+        {
+            try
+            {
+                var repo = gitExt.ActiveRepositories?.FirstOrDefault();
```

Comment:
This needs to be done on a background thread, see `TeamExplorerServiceHolder` for how it does this.

Reply:
I've fixed it so that `ActiveRepositories` is initialized by `VSGitExt` on a thread pool thread.